src/tail/session.rs

```diff
@@ -0,0 +1,118 @@
+use std::str;
+use std::thread;
+use std::time::Duration;
+
+use regex::Regex;
+use reqwest;
+use tokio::sync::oneshot::error::TryRecvError;
+use tokio::sync::oneshot::Receiver;
+use tokio::time;
+
+use cloudflare::endpoints::workers::{CreateTail, CreateTailParams, SendTailHeartbeat};
+use cloudflare::framework::HttpApiClientConfig;
+use cloudflare::framework::{async_api, async_api::ApiClient};
+
+use crate::http;
+use crate::settings::global_user::GlobalUser;
+use crate::settings::toml::Target;
+
+const KEEP_ALIVE_INTERVAL: u64 = 60;
+
+pub struct Session;
+
+impl Session {
+    pub async fn run(
+        target: Target,
+        user: GlobalUser,
+        mut rx: Receiver<()>,
+    ) -> Result<(), failure::Error> {
+        let client = http::cf_v4_api_client_async(&user, HttpApiClientConfig::default())?;
+
+        let url = get_tunnel_url().await?;
```

Comment:
TODO: i'd like tunnel to be in charge of its own url, and pass it in to this handler.